### DISCUSSION TITLE
incorrect URL when using reverse proxy

### DISCUSSION THREAD

**Discussion by paulens12** [firefly-iii/firefly-iii#5904] [Firefly III] [answered]
Excuse me if this is a silly question, but I'm new to this and I need help. I configured my reverse proxy as best as I could and I think the error is within Firefly itself, but please let me know if there's anything else that could cause it.

I used the Docker image to set everything up, and hid it behind an nginx host, proxying `https://domain.com/firefly/` to `http://domain.com:8080/`. As far as I can tell, the Nginx reverse proxy works fine because I can access individual pages without issues, such as `https://domain.com/firefly/login`.

The problem is that all the links and redirects in Firefly are broken. Starting with the default redirect, if I just open `https://domain.com/firefly/`, it redirects me to `https://domain.com/login` instead of `https://domain.com/firefly/login`.

When I open the login page, the sign in button doesn't do anything at all and the "I forgot my password" link leads to `http://domain.com/password/reset` instead of `http://domain.com/firefly/password/reset`.

Basically all links in Firefly are broken because they skip the `/firefly/` bit in the URL. This is very important to me because I run many different web applications on my server and I prefer using URL paths instead of different ports for everything.

**Comment by JC5** [comment]
Sorry, this isn't something that's supported unless you build the Docker image yourself.

**Reply by paulens12** [reply]
Oh, that's unexpected... Do you know what needs to be changed in the image to support it? I could open a PR to implement it as an environment variable.
I think that would be an extremely helpful addition. As it stands now, the image doesn't have built-in TLS and can't be used behind a reverse proxy, sounds to me like it was designed to be the only application running on a web server, which is besides the point of having a Docker image. With this addition it would be available to use in a microservice-like environment, which is what most people are aiming for when using Docker images.

**Reply by JC5** [reply]
>  I could open a PR to implement it as an environment variable.

No thanks.

> As it stands now, the image doesn't have built-in TLS and can't be used behind a reverse proxy

I don't know what you mean. Although it's true that the default web server runs on http by default (port 8080), it supports reverse proxies just fine. I'm running behind a reverse proxy right now, and with me many others.

**Reply by paulens12** [reply]
so how does your reverse proxy distinguish Firefly from other applications? I'm confused 🤔

**Reply by paulens12** [reply]
maybe I'm looking at the problem wrong. Instead of asking how to make Firefly respect the `/firefly` path I set up in Nginx, let me ask this: how do I put it behind a reverse proxy at all? How are you using it yourself, if you don't mind me asking?

**Reply by JC5** [reply]
Subdomains.

**Reply by JC5** [reply]
```
server {
	listen 443 ssl http2;
	listen [::]:443 ssl http2;
	root /var/www/html;
	index index.html;

	server_name ff.example.com;

	# example.com
	location / {
		proxy_set_header X-Forwarded-Host $host;
		proxy_set_header X-Forwarded-Server $host;
		proxy_set_header X-Forwarded-Proto $scheme;
		proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
		proxy_set_header Host $host;
		client_max_body_size 64M;
		proxy_read_timeout 300s;

		# pass to Docker container
		proxy_pass http://127.0.0.1:8080;

    }

	# SSL certificate.
	ssl_certificate /etc/openresty/tls/wildcard.crt;
	ssl_certificate_key /etc/openresty/tls/wildcard.key;
	include /etc/openresty/tls/config-firefly-iii.conf;
}
```

**Reply by paulens12** [reply]
oh ok, but then you need a DNS server running on your machine, don't you?

**Reply by JC5** [reply]
Or edit /etc/hosts or the Windows variant.

**Reply by paulens12** [reply]
LOL are you seriously suggesting editing the hosts file in each device I ever want to use Firefly from? Sorry but I can't put it any more mildly, that's ridiculous.

**Reply by JC5** [reply]
What you do is up to you, I'm just presenting an alternative. And since you said "on your machine", I presumed it was just the one, which doesn't seem to be the effort.

But please take my word for it, that subdomains is what most people use who self-host Firefly III. Usually though, they use Pi-Hole or similar solutions already and have a DNS server "in their pocket" already.

Since you ask, personally I don't host my Firefly III instance on the internet and I've added a local DNS entry to my own local DNS server. It works when you're on my WiFi, regardless of your device. The demo is hosted on Azure (sans Docker) and an instance for friends and family resides in a hosted VM. Both are "protected" by Cloudflare.

If you need anything else, let me know.

**Reply by paulens12** [reply]
oh, I should have made it clearer. By "your machine" I meant the machine running the Firefly software, not the client machine running the web browser.

I'm still considering building my own Docker image, do you have any pointers on what would need to be changed to support URL paths like in my example? Or does that require in-depth code changes? I'm not comfortable working with PHP so I would look for other options in that case.

I want to be able to use Firefly from my phone's LTE connection or whatever WiFi network it's connected to, which means I can't manually change the DNS server. The only way to make subdomains work would be by changing the domain from an "A" record at the registrar to an "NS" record and running a small DNS server of my own that would handle all the sub-domains. Sadly my registrar doesn't support wildcard domains.

**Reply by JC5** [reply]
Is this something you'd still like to know more about?

**Reply by paulens12** [reply]
Well, I do have a workaround so I'm good for now. It is quite flaky, though, and some functionality doesn't work, so I'll reiterate my question:

You said it's unsupported "unless you build the Docker image yourself" - does that mean there is some configuration option that could help me but isn't exposed in the Docker image? Or did you mean building the whole app myself with changes in the code?

Please let me know if there is a way to do this (URL subdirectory) without Docker. It would be much better than this ugly hack that barely works :)

Either way, thanks for sticking around with me, I know I can be annoying with my silly questions. I'm just new to self-hosting and still trying to figure out the best way to do things.

**Reply by JC5** [reply]
The Docker image runs on `http://localhost` but will generate `httpS://your.domain` if you send the right headers (see my nginx example).

That trick doesn't work for `httpS://your.domain/path` however. Either you use a reverse proxy to rewrite every single URL yourself or you create a new Docker image where it runs on `http://localhost/path`. This is something you can achieve by (for example) building a Docker image with the specific Apache `Alias`-directive to do this.

**Reply by paulens12** [reply]
Oh so if I just change the Apache config, your code will detect that? Looking at it from the frontend's perspective, the full URL is hardcoded in HTML and `/v1/jscript/variables`, so I just want to make sure we're on the same page.

**Reply by JC5** [reply]
The full URL is "hardcoded" in the html but in turn dynamically generated by the PHP code behind it. It works from subdirectories, so hosting it in a subdirectory is enough. That subdirectory will propogate through a reverse proxy.

**Comment by paulens12** [comment, marked as answer]
If anyone else comes here looking for a solution, I have a hacky workaround:
https://gist.github.com/paulens12/aeeed06e1772f65ac60a371066a217db
Build the docker image and run it alongside the official Firefly docker image.

This ugly hack is based on a less ugly (but still ugly) hack linked in the FAQ:
https://github.com/firefly-iii/firefly-iii/discussions/4892

I didn't want to switch everything I have from Nginx to Apache for this, so I just introduced another proxy layer.

**Reply by paulens12** [reply]
One nuisance is that you can't batch-edit anything using this workaround (or the original one from FAQ). So if I want to make a batch edit, I just log in over HTTP from my LAN, directly to the container instead of using a reverse proxy. Luckily that's not a feature I intend to use often :)

**Reply by paulens12** [reply]
Great news! I spent some more time debugging my issues and I managed to fix both batch edits and autofill (which was only partially working before). I've updated the gist accordingly.
